merge_command_test.go

```diff
@@ -197,158 +200,184 @@ var _ = TestWebhookHandler(func(context WebhookTestContext) {
-								mockMergeFailWithConflict()
-							})
-
-							It("retries and fails with a gateway error", func() {
-								handle()
-
-								// +1 because of the initial attempt
-								pullRequests.AssertNumberOfCalls(GinkgoT(), "Get", grh.MergeRetryLimit+1)
-								pullRequests.AssertNumberOfCalls(GinkgoT(), "Merge", grh.MergeRetryLimit+1)
-								Expect(responseRecorder.Code).To(Equal(http.StatusBadGateway))
-							})
-						})
-
-						Context("with merge succeeding with first retry", func() {
-							BeforeEach(func() {
-								mockMergeFailWithConflict().Once()
-
-								additionalCommitMessage := ""
-								pullRequests.
-									On(
-										"Merge",
-										repositoryOwner,
-										repositoryName,
-										issueNumber,
-										additionalCommitMessage,
-										noSquashOpts,
-									).
-									Return(&github.PullRequestMergeResult{
-										Merged: github.Bool(true),
-									}, nil, nil).
-									Once()
-							})
-
-							It("removes the 'merging' label from the PR after the merge", func() {
-								issues.
-									On("RemoveLabelForIssue", repositoryOwner, repositoryName, issueNumber, grh.MergingLabel).
-									Return(nil, nil, nil)
-
-								handle()
-								pullRequests.AssertNumberOfCalls(GinkgoT(), "Get", 2)
-								pullRequests.AssertNumberOfCalls(GinkgoT(), "Merge", 2)
-								Expect(responseRecorder.Code).To(Equal(http.StatusOK))
-							})
-						})
-					})
+var ItMergesPR = func(context WebhookTestContext, issueAuthor string, issueNumber int) {
+	var (
+		handle = context.Handle
+
+		responseRecorder *httptest.ResponseRecorder
+		pullRequests     *mocks.PullRequests
+		issues           *mocks.Issues
+	)
+	BeforeEach(func() {
+		responseRecorder = *context.ResponseRecorder
+		pullRequests = *context.PullRequests
+		issues = *context.Issues
+	})
 
-					Context("with merge failing, because PR not mergeable", func() {
-						BeforeEach(func() {
-							additionalCommitMessage := ""
-							resp := &http.Response{
-								StatusCode: http.StatusMethodNotAllowed,
-							}
-							pullRequests.
-								On(
-									"Merge",
-									repositoryOwner,
-									repositoryName,
-									issueNumber,
-									additionalCommitMessage,
-									noSquashOpts,
-								).
-								Return(nil, &github.Response{
-									Response: resp,
-								}, &github.ErrorResponse{
-									Response: resp,
-									Message:  "Pull Request is not mergeable",
-								}).
-								Once()
-						})
-
-						It("fails with a gateway error", func() {
-							handle()
-							Expect(responseRecorder.Code).To(Equal(http.StatusBadGateway))
-						})
-					})
+	Context("with merge failing with an unknown error", func() {
+		BeforeEach(func() {
+			additionalCommitMessage := ""
+			pullRequests.
+				On(
+					"Merge",
+					repositoryOwner,
+					repositoryName,
+					issueNumber,
+					additionalCommitMessage,
+					noSquashOpts,
+				).
+				Return(nil, nil, errors.New("an error")).
```

Comment:
This would add too much noise to the tests, I think. The pattern is almost uniformly the same throughout the GitHub lib's API. The 3 return values are: 1) result, 2) the lib's wrapper around `http.Response`, 3) error. The only exception that I've seen is the `RemoveLabelForIssue` endpoint which did cause some confusion, because I was returning 3 values from the mock element, but only expecting 2 in the code.

Creating names for all of the return values would add, on average, 2 lines to every mock setup. Creating a common `emptyResult` would be difficult, I think, because `nil`s also have types. The empty value for the 2nd return value could be shared between tests, because it's always of the same type.

I don't know. All in all I'm not currently convinced it's worth the noise + the effort.